fix: factor in monitor scaling for application library

pbui opened the pull request:
This addresses the problem where the application library is too large when a scaling factor is applied.
    
To fix this, we get the scale factor and apply to the height and width:
    
    monitorScale = 1/St.ThemeContext.get_for_stage(global.stage).scale_factor;
    height = Math.floor(monitorScale*monitor.height*.5 / 168) * 168
    width = Math.ceil(monitorScale*monitor.width*.6 / 168) * 168
    
Fixes #290.

pbui committed:
fix: factor in monitor scaling for application library

This addresses the problem where the application library is too large
when a scaling factor is applied.

To fix this, we get the scale factor and apply to the height and width:

    monitorScale = 1/St.ThemeContext.get_for_stage(global.stage).scale_factor;
    height = Math.floor(monitorScale*monitor.height*.5 / 168) * 168
    width = Math.ceil(monitorScale*monitor.width*.6 / 168) * 168

Fixes #290.

Whats-the-time commented:
Looks like we've fixed for some and broken others?  I have a 4k display @ 200% running perfectly until the latest updates broke it.  At 100% scaling it's fine (except with my old eyes, it's useless ;-)  ). 

![App_Launcher_200%](https://user-images.githubusercontent.com/97655442/149270381-4bd3c9d2-b4b4-4d85-ae51-93e1f23981f3.png)

pbui commented:
@Whats-the-time Is that screenshot with the fixes in this PR?

Whats-the-time commented:
> 
Just followed the linked issue on this page that led me to your fix proposed on reddit.  Applied the fix and it worked.  My apologies.  My search for the issue led me here and I assumed that this was part of the latest update that broke mine...it wasn't, in fact something else must have broken it, but your fix restored it.  Appreciate it.  Thanks for replying.

ids1024 commented:
I don't have a 4k monitor handy. Looking at this at 1440p with 200% scaling, this seems to fix the issue where the applications library tries to take more space than the screen.

But this just makes the icons the same size at 200% as they would be at 100%? Generally one would expect them to scale with the scale factor, and that may result in fairly small icons on something like a 4k 13 inch laptop.

mmstick commented:
Is the code currently monitoring scaling factor changes?

```js
St.ThemeContext.get_for_stage(global.stage)
    .connect('notify::scale-factor', () => this.update_scale());
```

pbui commented:
@mmstick It reads the `scale_factor` from the global stage each time the application library is resized (or made visible), so it shouldn't need to monitor it (unless the user changes the factor while the application library is open).

mmstick commented:
Makes sense then

pbui commented:
> But this just makes the icons the same size at 200% as they would be at 100%? Generally one would expect them to scale with the scale factor, and that may result in fairly small icons on something like a 4k 13 inch laptop.

Could you elaborate here?  Do you mean at 200% you'd want to see more icons (b/c they are smaller).  For instance, right now at 1080p I see 3 rows of 7 icons.  At larger resolutions, would you want to see more icons (ie. more rows or more icons per row)?

Currently for lower resolutions, I try to ensure that we maintain the same density (3x7).

For high resolutions (> 1080p), I tried to cap the iconScaling to 1 so that the width does not grow... with the hope that would allow for more icons per row (or more rows of icons).  I don't have a display > 1080p, so I cannot test if that actually works out or not.

If I could see a screenshot of what you are seeing at 1440p @ 100% and 200%, that might give me a better idea of what to correct.

Whats-the-time commented:
Hi Peter

Think this message may be intended for someone else?  I did not write
"But this just makes the icons the same size at 200% as they would be at
100%? Generally one would expect them to scale with the scale factor, and
that may result in fairly small icons on something like a 4k 13 inch
laptop."



On Thu, Jan 13, 2022 at 1:27 PM Peter Bui ***@***.***> wrote:

> But this just makes the icons the same size at 200% as they would be at
> 100%? Generally one would expect them to scale with the scale factor, and
> that may result in fairly small icons on something like a 4k 13 inch laptop.
>
> Could you elaborate here? Do you mean at 200% you'd want to see more icons
> (b/c they are smaller). For instance, right now at 1080p I see 3 rows of 7
> icons. At larger resolutions, would you want to see more icons (ie. more
> rows or more icons per row)?
>
> Currently for lower resolutions, I try to ensure that we maintain the same
> density (3x7).
>
> For high resolutions (> 1080p), I tried to cap the iconScaling to 1 so
> that the width does not grow... with the hope that would allow for more
> icons per row (or more rows of icons). I don't have a display > 1080p, so I
> cannot test if that actually works out or not.
>
> If I could see a screenshot of what you are seeing at 1440p @ 100% and
> 200%, that might give me a better idea of what to correct.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pop-os/cosmic/pull/277#issuecomment-1012529117>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXJBVEQW4KGVQEUQTJYPHNDUV47VLANCNFSM5KUDOVUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

pbui commented:
@Whats-the-time You are correct, that was in reply to @ids1024. 

Since you made a comment here, GitHub will send you an email with each new message, which is why you got that email.  You can unsubscribe from the notifications on the right hand side of the Pull Request if you wish to stop getting emails for this in the future.

Whats-the-time commented:
Thanks

ids1024 commented:
> I tried to cap the iconScaling to 1 so that the width does not grow.

Exactly. But that's not how scale factors generally work, and may be problematic at some DPIs.

In general, the idea of HiDPI scaling is that on a 4K monitor with a 2x scale factor, everything will be the exact same size as it would be on a 1080p monitor of the same size with a 1x scale, just sharper/higher resolution.

I don't currently have something to test it on, but by capping icon scaling to 1x, the icons would probably seem pretty small on a 13 inch laptop with a 4k display, for instance (things can be small on even a 13 inch 1080p laptop without fractional scaling).

Taking into account the scale factor when resizing the `appdisplay`, though, is an important fix. And it would be good to get that merged and released.

pbui commented:
> > I tried to cap the iconScaling to 1 so that the width does not grow.
> 
> Exactly. But that's not how scale factors generally work, and may be problematic at some DPIs.
> 
> In general, the idea of HiDPI scaling is that on a 4K monitor with a 2x scale factor, everything will be the exact same size as it would be on a 1080p monitor of the same size with a 1x scale, just sharper/higher resolution.

Ok.  Originally, I did not cap the iconScale to 1x to ensure we had the same density at every resolution and scale.  However, a user in #290 tried this version and said they would prefer to have more icons displayed:

https://github.com/pop-os/cosmic/issues/290#issuecomment-1009920753

If you would prefer that I don't cap the iconScale, then I can update the PR to remove that.

ids1024 commented:
> If you would prefer that I don't cap the iconScale, then I can update the PR to remove that.

Yeah, I think that's the right thing to do. I'm sure there's more room for improvement here, but capping the scale doesn't seem like the right solution.

pbui commented:
> > If you would prefer that I don't cap the iconScale, then I can update the PR to remove that.
> 
> Yeah, I think that's the right thing to do. I'm sure there's more room for improvement here, but capping the scale doesn't seem like the right solution.

Ok, I have updated the PR to no longer cap the iconScale factor.

Whats-the-time commented:
Thank you for all the suggestions.  Thought I'd add some screenshots of my 15" 4K display to show the progress (if that helps)

4k @ 200
![Screenshot3840x2160_200](https://user-images.githubusercontent.com/97655442/149426634-2a9cd7ee-830a-43c1-aa71-31e1002aebb1.png)

4k @ 100
![Screenshot3840x2160_100](https://user-images.githubusercontent.com/97655442/149426708-21e12ed2-cc4f-45e5-ad07-a8c2da4ad709.png)

1080 @ 100
![Screenshot1920x1080_100](https://user-images.githubusercontent.com/97655442/149426749-72059812-a7f8-4426-a7f7-d5ae9cd50491.png)

pbui commented:
I managed to borrow a friend's computer which has a 4K screen.  Here are screenshots with the latest patch at different fractional scaling:

4k @ 100%
![Screenshot from 2022-01-14 11-23-11](https://user-images.githubusercontent.com/3976244/149551325-6fdeff91-0c13-4784-beb7-a9b880a03859.png)

4k @ 125%
![Screenshot from 2022-01-14 11-23-29](https://user-images.githubusercontent.com/3976244/149551374-23d2b6d2-ca89-4a96-a23c-54b34b25e737.png)

4k @ 150%
![Screenshot from 2022-01-14 11-23-44](https://user-images.githubusercontent.com/3976244/149551421-35a4386a-22c7-4f1f-a34c-18aebb7350ed.png)

4k @ 175%
![Screenshot from 2022-01-14 11-23-55](https://user-images.githubusercontent.com/3976244/149551499-0b6227ff-639e-48e5-b6b9-c14cbfe783cd.png)

4k @ 200%
![Screenshot from 2022-01-14 11-24-09](https://user-images.githubusercontent.com/3976244/149551549-5f0c2da1-2591-4e62-8209-4a402f281c28.png)

As can be seen, the icon density now remains the same, regardless of the fractional setting.

ids1024 commented:
That looks good at 4k/200%. Though normally everything should scale with the scale factor so the app library and icons should be small at 4k/100%.

It seems if I set it through xrandr, I can get my 1440p monitor to input a 4k/30 signal. I think I'll some things testing with that.

ids1024 commented:
I think the behavior better matches expected behavior just not resizing icons, and only changing the `_scrollView` size. Then Gnome Shell just scales them by the scale factor. At 4k/100%:

![Screenshot from 2022-01-14 10-15-07](https://user-images.githubusercontent.com/2263150/149565184-6a86e619-1bed-44de-8b1c-111fd9374856.png)

The icons are tiny, but that is expected since everything else is.

pbui commented:
The reason for scaling the icons at all was for smaller resolutions (ie. < 1080).  Currently with smaller resolutions, we would have less density and thus it shows less icons b/c of their size.

Would you prefer we not resize the icons at all, at the cost of the smaller resolution use case?

ids1024 commented:
Ah, sorry about the confusion. I didn't read the description for the PR carefully enough, and didn't realize the icon scaling was meant to help with lower resolutions.

Maybe it would be better it merge without the icon scaling, to fix the current bug on HiDPI systems, and later make an improvement for lower resolution displays (that handles search results as well, and doesn't impact behavior on higher resolutions).

pbui commented:
@ids1024 Yep.  I've updated the PR to only fix the application library window size and removed the icon scaling part.

I think for icon scaling, we should consider having a user option that let's them specify the size rather than trying to automatically scale based on the resolution (similar to the dock icon size).

ids1024 commented:
Yeah, to some extent it's also a matter of user preference, and it wouldn't be out of place to have such an option given the dock settings.

jacobgkau reviewed (approved):
Confirmed this fixes https://github.com/pop-os/cosmic/issues/290.

Tested the app grid in the following configs:
- [X] lemp10 (LoDPI)
- [X] addw2 (HiDPI)
- [X] lemp10 connected to LoDPI external display (opened app grid on each display)
- [ ] addw2 connected to LoDPI external display (opened app grid on each display)
- [X] lemp10 connected to HiDPI external display (opened app grid on each display)
- [X] addw2 connected to HiDPI external display (opened app grid on each display)

On a 4K display being run at LoDPI (no scaling up), additional icons are displayed as shown in https://github.com/pop-os/cosmic/pull/277#issuecomment-1013351496, which says that's expected. This also matches current behavior.

When I connected addw2 (HiDPI 4K) to a LoDPI 1080p external display, the scaling was wrong when opening the app grid on the LoDPI display:

![Screenshot from 2022-01-14 14-54-57](https://user-images.githubusercontent.com/7199422/149590844-a28c076e-e36d-4e63-a0b6-e3325d8de620.jpg)

However, the LoDPI display was previously affected by https://github.com/pop-os/cosmic/issues/290 as shown below, so that config isn't a regression (the overall positioning is still an improvement):

![Screenshot from 2022-01-14 14-55-50](https://user-images.githubusercontent.com/7199422/149590850-d07894d0-be3d-494e-b736-ae6e6126cffb.jpg)
